{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    },
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Name: ultralytics\n",
      "Version: 8.3.61\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "--- Logging error ---\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\quocd\\AppData\\Local\\uv\\cache\\archive-v0\\z8lkNhNwG6XGWGBlplZML\\Lib\\site-packages\\pip\\_internal\\utils\\logging.py\", line 177, in emit\n",
      "    self.console.print(renderable, overflow=\"ignore\", crop=False, style=style)\n",
      "  File \"C:\\Users\\quocd\\AppData\\Local\\uv\\cache\\archive-v0\\z8lkNhNwG6XGWGBlplZML\\Lib\\site-packages\\pip\\_vendor\\rich\\console.py\", line 1673, in print\n",
      "    with self:\n",
      "         ^^^^\n",
      "  File \"C:\\Users\\quocd\\AppData\\Local\\uv\\cache\\archive-v0\\z8lkNhNwG6XGWGBlplZML\\Lib\\site-packages\\pip\\_vendor\\rich\\console.py\", line 865, in __exit__\n",
      "    self._exit_buffer()\n",
      "  File \"C:\\Users\\quocd\\AppData\\Local\\uv\\cache\\archive-v0\\z8lkNhNwG6XGWGBlplZML\\Lib\\site-packages\\pip\\_vendor\\rich\\console.py\", line 823, in _exit_buffer\n",
      "    self._check_buffer()\n",
      "  File \"C:\\Users\\quocd\\AppData\\Local\\uv\\cache\\archive-v0\\z8lkNhNwG6XGWGBlplZML\\Lib\\site-packages\\pip\\_vendor\\rich\\console.py\", line 2027, in _check_buffer\n",
      "    legacy_windows_render(buffer, LegacyWindowsTerm(self.file))\n",
      "  File \"C:\\Users\\quocd\\AppData\\Local\\uv\\cache\\archive-v0\\z8lkNhNwG6XGWGBlplZML\\Lib\\site-packages\\pip\\_vendor\\rich\\_windows_renderer.py\", line 19, in legacy_windows_render\n",
      "    term.write_text(text)\n",
      "  File \"C:\\Users\\quocd\\AppData\\Local\\uv\\cache\\archive-v0\\z8lkNhNwG6XGWGBlplZML\\Lib\\site-packages\\pip\\_vendor\\rich\\_win32_console.py\", line 403, in write_text\n",
      "    self.write(text)\n",
      "  File \"C:\\Users\\quocd\\AppData\\Roaming\\uv\\python\\cpython-3.12.8-windows-x86_64-none\\Lib\\encodings\\cp1252.py\", line 19, in encode\n",
      "    return codecs.charmap_encode(input,self.errors,encoding_table)[0]\n",
      "           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "UnicodeEncodeError: 'charmap' codec can't encode character '\\U0001f680' in position 26: character maps to <undefined>\n",
      "Call stack:\n",
      "  File \"<frozen runpy>\", line 198, in _run_module_as_main\n",
      "  File \"<frozen runpy>\", line 88, in _run_code\n",
      "  File \"C:\\Users\\quocd\\AppData\\Local\\uv\\cache\\archive-v0\\z8lkNhNwG6XGWGBlplZML\\Scripts\\pip.exe\\__main__.py\", line 8, in <module>\n",
      "    sys.exit(main())\n",
      "  File \"C:\\Users\\quocd\\AppData\\Local\\uv\\cache\\archive-v0\\z8lkNhNwG6XGWGBlplZML\\Lib\\site-packages\\pip\\_internal\\cli\\main.py\", line 80, in main\n",
      "    return command.main(cmd_args)\n",
      "  File \"C:\\Users\\quocd\\AppData\\Local\\uv\\cache\\archive-v0\\z8lkNhNwG6XGWGBlplZML\\Lib\\site-packages\\pip\\_internal\\cli\\base_command.py\", line 157, in main\n",
      "    return self._main(args)\n",
      "  File \"C:\\Users\\quocd\\AppData\\Local\\uv\\cache\\archive-v0\\z8lkNhNwG6XGWGBlplZML\\Lib\\site-packages\\pip\\_internal\\cli\\base_command.py\", line 231, in _main\n",
      "    return self._run_wrapper(level_number, options, args)\n",
      "  File \"C:\\Users\\quocd\\AppData\\Local\\uv\\cache\\archive-v0\\z8lkNhNwG6XGWGBlplZML\\Lib\\site-packages\\pip\\_internal\\cli\\base_command.py\", line 105, in _run_wrapper\n",
      "    status = _inner_run()\n",
      "  File \"C:\\Users\\quocd\\AppData\\Local\\uv\\cache\\archive-v0\\z8lkNhNwG6XGWGBlplZML\\Lib\\site-packages\\pip\\_internal\\cli\\base_command.py\", line 96, in _inner_run\n",
      "    return self.run(options, args)\n",
      "  File \"C:\\Users\\quocd\\AppData\\Local\\uv\\cache\\archive-v0\\z8lkNhNwG6XGWGBlplZML\\Lib\\site-packages\\pip\\_internal\\commands\\show.py\", line 46, in run\n",
      "    if not print_results(\n",
      "  File \"C:\\Users\\quocd\\AppData\\Local\\uv\\cache\\archive-v0\\z8lkNhNwG6XGWGBlplZML\\Lib\\site-packages\\pip\\_internal\\commands\\show.py\", line 185, in print_results\n",
      "    write_output(\"Summary: %s\", dist.summary)\n",
      "  File \"C:\\Users\\quocd\\AppData\\Local\\uv\\cache\\archive-v0\\z8lkNhNwG6XGWGBlplZML\\Lib\\site-packages\\pip\\_internal\\utils\\misc.py\", line 374, in write_output\n",
      "    logger.info(msg, *args)\n",
      "  File \"C:\\Users\\quocd\\AppData\\Roaming\\uv\\python\\cpython-3.12.8-windows-x86_64-none\\Lib\\logging\\__init__.py\", line 1539, in info\n",
      "    self._log(INFO, msg, args, **kwargs)\n",
      "  File \"C:\\Users\\quocd\\AppData\\Roaming\\uv\\python\\cpython-3.12.8-windows-x86_64-none\\Lib\\logging\\__init__.py\", line 1684, in _log\n",
      "    self.handle(record)\n",
      "  File \"C:\\Users\\quocd\\AppData\\Roaming\\uv\\python\\cpython-3.12.8-windows-x86_64-none\\Lib\\logging\\__init__.py\", line 1700, in handle\n",
      "    self.callHandlers(record)\n",
      "  File \"C:\\Users\\quocd\\AppData\\Roaming\\uv\\python\\cpython-3.12.8-windows-x86_64-none\\Lib\\logging\\__init__.py\", line 1762, in callHandlers\n",
      "    hdlr.handle(record)\n",
      "  File \"C:\\Users\\quocd\\AppData\\Roaming\\uv\\python\\cpython-3.12.8-windows-x86_64-none\\Lib\\logging\\__init__.py\", line 1028, in handle\n",
      "    self.emit(record)\n",
      "  File \"C:\\Users\\quocd\\AppData\\Local\\uv\\cache\\archive-v0\\z8lkNhNwG6XGWGBlplZML\\Lib\\site-packages\\pip\\_internal\\utils\\logging.py\", line 179, in emit\n",
      "    self.handleError(record)\n",
      "Message: 'Summary: %s'\n",
      "Arguments: ('Ultralytics YOLO \\U0001f680 for SOTA object detection, multi-object tracking, instance segmentation, pose estimation and image classification.',)\n",
      "--- Logging error ---\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\quocd\\AppData\\Local\\uv\\cache\\archive-v0\\z8lkNhNwG6XGWGBlplZML\\Lib\\site-packages\\pip\\_internal\\utils\\logging.py\", line 177, in emit\n",
      "    self.console.print(renderable, overflow=\"ignore\", crop=False, style=style)\n",
      "  File \"C:\\Users\\quocd\\AppData\\Local\\uv\\cache\\archive-v0\\z8lkNhNwG6XGWGBlplZML\\Lib\\site-packages\\pip\\_vendor\\rich\\console.py\", line 1673, in print\n",
      "    with self:\n",
      "         ^^^^\n",
      "  File \"C:\\Users\\quocd\\AppData\\Local\\uv\\cache\\archive-v0\\z8lkNhNwG6XGWGBlplZML\\Lib\\site-packages\\pip\\_vendor\\rich\\console.py\", line 865, in __exit__\n",
      "    self._exit_buffer()\n",
      "  File \"C:\\Users\\quocd\\AppData\\Local\\uv\\cache\\archive-v0\\z8lkNhNwG6XGWGBlplZML\\Lib\\site-packages\\pip\\_vendor\\rich\\console.py\", line 823, in _exit_buffer\n",
      "    self._check_buffer()\n",
      "  File \"C:\\Users\\quocd\\AppData\\Local\\uv\\cache\\archive-v0\\z8lkNhNwG6XGWGBlplZML\\Lib\\site-packages\\pip\\_vendor\\rich\\console.py\", line 2027, in _check_buffer\n",
      "    legacy_windows_render(buffer, LegacyWindowsTerm(self.file))\n",
      "  File \"C:\\Users\\quocd\\AppData\\Local\\uv\\cache\\archive-v0\\z8lkNhNwG6XGWGBlplZML\\Lib\\site-packages\\pip\\_vendor\\rich\\_windows_renderer.py\", line 19, in legacy_windows_render\n",
      "    term.write_text(text)\n",
      "  File \"C:\\Users\\quocd\\AppData\\Local\\uv\\cache\\archive-v0\\z8lkNhNwG6XGWGBlplZML\\Lib\\site-packages\\pip\\_vendor\\rich\\_win32_console.py\", line 403, in write_text\n",
      "    self.write(text)\n",
      "  File \"C:\\Users\\quocd\\AppData\\Roaming\\uv\\python\\cpython-3.12.8-windows-x86_64-none\\Lib\\encodings\\cp1252.py\", line 19, in encode\n",
      "    return codecs.charmap_encode(input,self.errors,encoding_table)[0]\n",
      "           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "UnicodeEncodeError: 'charmap' codec can't encode character '\\U0001f680' in position 26: character maps to <undefined>\n",
      "Call stack:\n",
      "  File \"<frozen runpy>\", line 198, in _run_module_as_main\n",
      "  File \"<frozen runpy>\", line 88, in _run_code\n",
      "  File \"C:\\Users\\quocd\\AppData\\Local\\uv\\cache\\archive-v0\\z8lkNhNwG6XGWGBlplZML\\Scripts\\pip.exe\\__main__.py\", line 8, in <module>\n",
      "    sys.exit(main())\n",
      "  File \"C:\\Users\\quocd\\AppData\\Local\\uv\\cache\\archive-v0\\z8lkNhNwG6XGWGBlplZML\\Lib\\site-packages\\pip\\_internal\\cli\\main.py\", line 80, in main\n",
      "    return command.main(cmd_args)\n",
      "  File \"C:\\Users\\quocd\\AppData\\Local\\uv\\cache\\archive-v0\\z8lkNhNwG6XGWGBlplZML\\Lib\\site-packages\\pip\\_internal\\cli\\base_command.py\", line 157, in main\n",
      "    return self._main(args)\n",
      "  File \"C:\\Users\\quocd\\AppData\\Local\\uv\\cache\\archive-v0\\z8lkNhNwG6XGWGBlplZML\\Lib\\site-packages\\pip\\_internal\\cli\\base_command.py\", line 231, in _main\n",
      "    return self._run_wrapper(level_number, options, args)\n",
      "  File \"C:\\Users\\quocd\\AppData\\Local\\uv\\cache\\archive-v0\\z8lkNhNwG6XGWGBlplZML\\Lib\\site-packages\\pip\\_internal\\cli\\base_command.py\", line 105, in _run_wrapper\n",
      "    status = _inner_run()\n",
      "  File \"C:\\Users\\quocd\\AppData\\Local\\uv\\cache\\archive-v0\\z8lkNhNwG6XGWGBlplZML\\Lib\\site-packages\\pip\\_internal\\cli\\base_command.py\", line 96, in _inner_run\n",
      "    return self.run(options, args)\n",
      "  File \"C:\\Users\\quocd\\AppData\\Local\\uv\\cache\\archive-v0\\z8lkNhNwG6XGWGBlplZML\\Lib\\site-packages\\pip\\_internal\\commands\\show.py\", line 46, in run\n",
      "    if not print_results(\n",
      "  File \"C:\\Users\\quocd\\AppData\\Local\\uv\\cache\\archive-v0\\z8lkNhNwG6XGWGBlplZML\\Lib\\site-packages\\pip\\_internal\\commands\\show.py\", line 186, in print_results\n",
      "    write_output(\"Home-page: %s\", dist.homepage)\n",
      "  File \"C:\\Users\\quocd\\AppData\\Local\\uv\\cache\\archive-v0\\z8lkNhNwG6XGWGBlplZML\\Lib\\site-packages\\pip\\_internal\\utils\\misc.py\", line 374, in write_output\n",
      "    logger.info(msg, *args)\n",
      "  File \"C:\\Users\\quocd\\AppData\\Roaming\\uv\\python\\cpython-3.12.8-windows-x86_64-none\\Lib\\logging\\__init__.py\", line 1539, in info\n",
      "    self._log(INFO, msg, args, **kwargs)\n",
      "  File \"C:\\Users\\quocd\\AppData\\Roaming\\uv\\python\\cpython-3.12.8-windows-x86_64-none\\Lib\\logging\\__init__.py\", line 1684, in _log\n",
      "    self.handle(record)\n",
      "  File \"C:\\Users\\quocd\\AppData\\Roaming\\uv\\python\\cpython-3.12.8-windows-x86_64-none\\Lib\\logging\\__init__.py\", line 1700, in handle\n",
      "    self.callHandlers(record)\n",
      "  File \"C:\\Users\\quocd\\AppData\\Roaming\\uv\\python\\cpython-3.12.8-windows-x86_64-none\\Lib\\logging\\__init__.py\", line 1762, in callHandlers\n",
      "    hdlr.handle(record)\n",
      "  File \"C:\\Users\\quocd\\AppData\\Roaming\\uv\\python\\cpython-3.12.8-windows-x86_64-none\\Lib\\logging\\__init__.py\", line 1028, in handle\n",
      "    self.emit(record)\n",
      "  File \"C:\\Users\\quocd\\AppData\\Local\\uv\\cache\\archive-v0\\z8lkNhNwG6XGWGBlplZML\\Lib\\site-packages\\pip\\_internal\\utils\\logging.py\", line 179, in emit\n",
      "    self.handleError(record)\n",
      "Message: 'Home-page: %s'\n",
      "Arguments: ('https://ultralytics.com',)\n",
      "--- Logging error ---\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\quocd\\AppData\\Local\\uv\\cache\\archive-v0\\z8lkNhNwG6XGWGBlplZML\\Lib\\site-packages\\pip\\_internal\\utils\\logging.py\", line 177, in emit\n",
      "    self.console.print(renderable, overflow=\"ignore\", crop=False, style=style)\n",
      "  File \"C:\\Users\\quocd\\AppData\\Local\\uv\\cache\\archive-v0\\z8lkNhNwG6XGWGBlplZML\\Lib\\site-packages\\pip\\_vendor\\rich\\console.py\", line 1673, in print\n",
      "    with self:\n",
      "         ^^^^\n",
      "  File \"C:\\Users\\quocd\\AppData\\Local\\uv\\cache\\archive-v0\\z8lkNhNwG6XGWGBlplZML\\Lib\\site-packages\\pip\\_vendor\\rich\\console.py\", line 865, in __exit__\n",
      "    self._exit_buffer()\n",
      "  File \"C:\\Users\\quocd\\AppData\\Local\\uv\\cache\\archive-v0\\z8lkNhNwG6XGWGBlplZML\\Lib\\site-packages\\pip\\_vendor\\rich\\console.py\", line 823, in _exit_buffer\n",
      "    self._check_buffer()\n",
      "  File \"C:\\Users\\quocd\\AppData\\Local\\uv\\cache\\archive-v0\\z8lkNhNwG6XGWGBlplZML\\Lib\\site-packages\\pip\\_vendor\\rich\\console.py\", line 2027, in _check_buffer\n",
      "    legacy_windows_render(buffer, LegacyWindowsTerm(self.file))\n",
      "  File \"C:\\Users\\quocd\\AppData\\Local\\uv\\cache\\archive-v0\\z8lkNhNwG6XGWGBlplZML\\Lib\\site-packages\\pip\\_vendor\\rich\\_windows_renderer.py\", line 19, in legacy_windows_render\n",
      "    term.write_text(text)\n",
      "  File \"C:\\Users\\quocd\\AppData\\Local\\uv\\cache\\archive-v0\\z8lkNhNwG6XGWGBlplZML\\Lib\\site-packages\\pip\\_vendor\\rich\\_win32_console.py\", line 403, in write_text\n",
      "    self.write(text)\n",
      "  File \"C:\\Users\\quocd\\AppData\\Roaming\\uv\\python\\cpython-3.12.8-windows-x86_64-none\\Lib\\encodings\\cp1252.py\", line 19, in encode\n",
      "    return codecs.charmap_encode(input,self.errors,encoding_table)[0]\n",
      "           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "UnicodeEncodeError: 'charmap' codec can't encode character '\\U0001f680' in position 26: character maps to <undefined>\n",
      "Call stack:\n",
      "  File \"<frozen runpy>\", line 198, in _run_module_as_main\n",
      "  File \"<frozen runpy>\", line 88, in _run_code\n",
      "  File \"C:\\Users\\quocd\\AppData\\Local\\uv\\cache\\archive-v0\\z8lkNhNwG6XGWGBlplZML\\Scripts\\pip.exe\\__main__.py\", line 8, in <module>\n",
      "    sys.exit(main())\n",
      "  File \"C:\\Users\\quocd\\AppData\\Local\\uv\\cache\\archive-v0\\z8lkNhNwG6XGWGBlplZML\\Lib\\site-packages\\pip\\_internal\\cli\\main.py\", line 80, in main\n",
      "    return command.main(cmd_args)\n",
      "  File \"C:\\Users\\quocd\\AppData\\Local\\uv\\cache\\archive-v0\\z8lkNhNwG6XGWGBlplZML\\Lib\\site-packages\\pip\\_internal\\cli\\base_command.py\", line 157, in main\n",
      "    return self._main(args)\n",
      "  File \"C:\\Users\\quocd\\AppData\\Local\\uv\\cache\\archive-v0\\z8lkNhNwG6XGWGBlplZML\\Lib\\site-packages\\pip\\_internal\\cli\\base_command.py\", line 231, in _main\n",
      "    return self._run_wrapper(level_number, options, args)\n",
      "  File \"C:\\Users\\quocd\\AppData\\Local\\uv\\cache\\archive-v0\\z8lkNhNwG6XGWGBlplZML\\Lib\\site-packages\\pip\\_internal\\cli\\base_command.py\", line 105, in _run_wrapper\n",
      "    status = _inner_run()\n",
      "  File \"C:\\Users\\quocd\\AppData\\Local\\uv\\cache\\archive-v0\\z8lkNhNwG6XGWGBlplZML\\Lib\\site-packages\\pip\\_internal\\cli\\base_command.py\", line 96, in _inner_run\n",
      "    return self.run(options, args)\n",
      "  File \"C:\\Users\\quocd\\AppData\\Local\\uv\\cache\\archive-v0\\z8lkNhNwG6XGWGBlplZML\\Lib\\site-packages\\pip\\_internal\\commands\\show.py\", line 46, in run\n",
      "    if not print_results(\n",
      "  File \"C:\\Users\\quocd\\AppData\\Local\\uv\\cache\\archive-v0\\z8lkNhNwG6XGWGBlplZML\\Lib\\site-packages\\pip\\_internal\\commands\\show.py\", line 187, in print_results\n",
      "    write_output(\"Author: %s\", dist.author)\n",
      "  File \"C:\\Users\\quocd\\AppData\\Local\\uv\\cache\\archive-v0\\z8lkNhNwG6XGWGBlplZML\\Lib\\site-packages\\pip\\_internal\\utils\\misc.py\", line 374, in write_output\n",
      "    logger.info(msg, *args)\n",
      "  File \"C:\\Users\\quocd\\AppData\\Roaming\\uv\\python\\cpython-3.12.8-windows-x86_64-none\\Lib\\logging\\__init__.py\", line 1539, in info\n",
      "    self._log(INFO, msg, args, **kwargs)\n",
      "  File \"C:\\Users\\quocd\\AppData\\Roaming\\uv\\python\\cpython-3.12.8-windows-x86_64-none\\Lib\\logging\\__init__.py\", line 1684, in _log\n",
      "    self.handle(record)\n",
      "  File \"C:\\Users\\quocd\\AppData\\Roaming\\uv\\python\\cpython-3.12.8-windows-x86_64-none\\Lib\\logging\\__init__.py\", line 1700, in handle\n",
      "    self.callHandlers(record)\n",
      "  File \"C:\\Users\\quocd\\AppData\\Roaming\\uv\\python\\cpython-3.12.8-windows-x86_64-none\\Lib\\logging\\__init__.py\", line 1762, in callHandlers\n",
      "    hdlr.handle(record)\n",
      "  File \"C:\\Users\\quocd\\AppData\\Roaming\\uv\\python\\cpython-3.12.8-windows-x86_64-none\\Lib\\logging\\__init__.py\", line 1028, in handle\n",
      "    self.emit(record)\n",
      "  File \"C:\\Users\\quocd\\AppData\\Local\\uv\\cache\\archive-v0\\z8lkNhNwG6XGWGBlplZML\\Lib\\site-packages\\pip\\_internal\\utils\\logging.py\", line 179, in emit\n",
      "    self.handleError(record)\n",
      "Message: 'Author: %s'\n",
      "Arguments: ('',)\n",
      "--- Logging error ---\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\quocd\\AppData\\Local\\uv\\cache\\archive-v0\\z8lkNhNwG6XGWGBlplZML\\Lib\\site-packages\\pip\\_internal\\utils\\logging.py\", line 177, in emit\n",
      "    self.console.print(renderable, overflow=\"ignore\", crop=False, style=style)\n",
      "  File \"C:\\Users\\quocd\\AppData\\Local\\uv\\cache\\archive-v0\\z8lkNhNwG6XGWGBlplZML\\Lib\\site-packages\\pip\\_vendor\\rich\\console.py\", line 1673, in print\n",
      "    with self:\n",
      "         ^^^^\n",
      "  File \"C:\\Users\\quocd\\AppData\\Local\\uv\\cache\\archive-v0\\z8lkNhNwG6XGWGBlplZML\\Lib\\site-packages\\pip\\_vendor\\rich\\console.py\", line 865, in __exit__\n",
      "    self._exit_buffer()\n",
      "  File \"C:\\Users\\quocd\\AppData\\Local\\uv\\cache\\archive-v0\\z8lkNhNwG6XGWGBlplZML\\Lib\\site-packages\\pip\\_vendor\\rich\\console.py\", line 823, in _exit_buffer\n",
      "    self._check_buffer()\n",
      "  File \"C:\\Users\\quocd\\AppData\\Local\\uv\\cache\\archive-v0\\z8lkNhNwG6XGWGBlplZML\\Lib\\site-packages\\pip\\_vendor\\rich\\console.py\", line 2027, in _check_buffer\n",
      "    legacy_windows_render(buffer, LegacyWindowsTerm(self.file))\n",
      "  File \"C:\\Users\\quocd\\AppData\\Local\\uv\\cache\\archive-v0\\z8lkNhNwG6XGWGBlplZML\\Lib\\site-packages\\pip\\_vendor\\rich\\_windows_renderer.py\", line 19, in legacy_windows_render\n",
      "    term.write_text(text)\n",
      "  File \"C:\\Users\\quocd\\AppData\\Local\\uv\\cache\\archive-v0\\z8lkNhNwG6XGWGBlplZML\\Lib\\site-packages\\pip\\_vendor\\rich\\_win32_console.py\", line 403, in write_text\n",
      "    self.write(text)\n",
      "  File \"C:\\Users\\quocd\\AppData\\Roaming\\uv\\python\\cpython-3.12.8-windows-x86_64-none\\Lib\\encodings\\cp1252.py\", line 19, in encode\n",
      "    return codecs.charmap_encode(input,self.errors,encoding_table)[0]\n",
      "           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "UnicodeEncodeError: 'charmap' codec can't encode character '\\U0001f680' in position 26: character maps to <undefined>\n",
      "Call stack:\n",
      "  File \"<frozen runpy>\", line 198, in _run_module_as_main\n",
      "  File \"<frozen runpy>\", line 88, in _run_code\n",
      "  File \"C:\\Users\\quocd\\AppData\\Local\\uv\\cache\\archive-v0\\z8lkNhNwG6XGWGBlplZML\\Scripts\\pip.exe\\__main__.py\", line 8, in <module>\n",
      "    sys.exit(main())\n",
      "  File \"C:\\Users\\quocd\\AppData\\Local\\uv\\cache\\archive-v0\\z8lkNhNwG6XGWGBlplZML\\Lib\\site-packages\\pip\\_internal\\cli\\main.py\", line 80, in main\n",
      "    return command.main(cmd_args)\n",
      "  File \"C:\\Users\\quocd\\AppData\\Local\\uv\\cache\\archive-v0\\z8lkNhNwG6XGWGBlplZML\\Lib\\site-packages\\pip\\_internal\\cli\\base_command.py\", line 157, in main\n",
      "    return self._main(args)\n",
      "  File \"C:\\Users\\quocd\\AppData\\Local\\uv\\cache\\archive-v0\\z8lkNhNwG6XGWGBlplZML\\Lib\\site-packages\\pip\\_internal\\cli\\base_command.py\", line 231, in _main\n",
      "    return self._run_wrapper(level_number, options, args)\n",
      "  File \"C:\\Users\\quocd\\AppData\\Local\\uv\\cache\\archive-v0\\z8lkNhNwG6XGWGBlplZML\\Lib\\site-packages\\pip\\_internal\\cli\\base_command.py\", line 105, in _run_wrapper\n",
      "    status = _inner_run()\n",
      "  File \"C:\\Users\\quocd\\AppData\\Local\\uv\\cache\\archive-v0\\z8lkNhNwG6XGWGBlplZML\\Lib\\site-packages\\pip\\_internal\\cli\\base_command.py\", line 96, in _inner_run\n",
      "    return self.run(options, args)\n",
      "  File \"C:\\Users\\quocd\\AppData\\Local\\uv\\cache\\archive-v0\\z8lkNhNwG6XGWGBlplZML\\Lib\\site-packages\\pip\\_internal\\commands\\show.py\", line 46, in run\n",
      "    if not print_results(\n",
      "  File \"C:\\Users\\quocd\\AppData\\Local\\uv\\cache\\archive-v0\\z8lkNhNwG6XGWGBlplZML\\Lib\\site-packages\\pip\\_internal\\commands\\show.py\", line 188, in print_results\n",
      "    write_output(\"Author-email: %s\", dist.author_email)\n",
      "  File \"C:\\Users\\quocd\\AppData\\Local\\uv\\cache\\archive-v0\\z8lkNhNwG6XGWGBlplZML\\Lib\\site-packages\\pip\\_internal\\utils\\misc.py\", line 374, in write_output\n",
      "    logger.info(msg, *args)\n",
      "  File \"C:\\Users\\quocd\\AppData\\Roaming\\uv\\python\\cpython-3.12.8-windows-x86_64-none\\Lib\\logging\\__init__.py\", line 1539, in info\n",
      "    self._log(INFO, msg, args, **kwargs)\n",
      "  File \"C:\\Users\\quocd\\AppData\\Roaming\\uv\\python\\cpython-3.12.8-windows-x86_64-none\\Lib\\logging\\__init__.py\", line 1684, in _log\n",
      "    self.handle(record)\n",
      "  File \"C:\\Users\\quocd\\AppData\\Roaming\\uv\\python\\cpython-3.12.8-windows-x86_64-none\\Lib\\logging\\__init__.py\", line 1700, in handle\n",
      "    self.callHandlers(record)\n",
      "  File \"C:\\Users\\quocd\\AppData\\Roaming\\uv\\python\\cpython-3.12.8-windows-x86_64-none\\Lib\\logging\\__init__.py\", line 1762, in callHandlers\n",
      "    hdlr.handle(record)\n",
      "  File \"C:\\Users\\quocd\\AppData\\Roaming\\uv\\python\\cpython-3.12.8-windows-x86_64-none\\Lib\\logging\\__init__.py\", line 1028, in handle\n",
      "    self.emit(record)\n",
      "  File \"C:\\Users\\quocd\\AppData\\Local\\uv\\cache\\archive-v0\\z8lkNhNwG6XGWGBlplZML\\Lib\\site-packages\\pip\\_internal\\utils\\logging.py\", line 179, in emit\n",
      "    self.handleError(record)\n",
      "Message: 'Author-email: %s'\n",
      "Arguments: ('Glenn Jocher <glenn.jocher@ultralytics.com>, Jing Qiu <jing.qiu@ultralytics.com>',)\n",
      "--- Logging error ---\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\quocd\\AppData\\Local\\uv\\cache\\archive-v0\\z8lkNhNwG6XGWGBlplZML\\Lib\\site-packages\\pip\\_internal\\utils\\logging.py\", line 177, in emit\n",
      "    self.console.print(renderable, overflow=\"ignore\", crop=False, style=style)\n",
      "  File \"C:\\Users\\quocd\\AppData\\Local\\uv\\cache\\archive-v0\\z8lkNhNwG6XGWGBlplZML\\Lib\\site-packages\\pip\\_vendor\\rich\\console.py\", line 1673, in print\n",
      "    with self:\n",
      "         ^^^^\n",
      "  File \"C:\\Users\\quocd\\AppData\\Local\\uv\\cache\\archive-v0\\z8lkNhNwG6XGWGBlplZML\\Lib\\site-packages\\pip\\_vendor\\rich\\console.py\", line 865, in __exit__\n",
      "    self._exit_buffer()\n",
      "  File \"C:\\Users\\quocd\\AppData\\Local\\uv\\cache\\archive-v0\\z8lkNhNwG6XGWGBlplZML\\Lib\\site-packages\\pip\\_vendor\\rich\\console.py\", line 823, in _exit_buffer\n",
      "    self._check_buffer()\n",
      "  File \"C:\\Users\\quocd\\AppData\\Local\\uv\\cache\\archive-v0\\z8lkNhNwG6XGWGBlplZML\\Lib\\site-packages\\pip\\_vendor\\rich\\console.py\", line 2027, in _check_buffer\n",
      "    legacy_windows_render(buffer, LegacyWindowsTerm(self.file))\n",
      "  File \"C:\\Users\\quocd\\AppData\\Local\\uv\\cache\\archive-v0\\z8lkNhNwG6XGWGBlplZML\\Lib\\site-packages\\pip\\_vendor\\rich\\_windows_renderer.py\", line 19, in legacy_windows_render\n",
      "    term.write_text(text)\n",
      "  File \"C:\\Users\\quocd\\AppData\\Local\\uv\\cache\\archive-v0\\z8lkNhNwG6XGWGBlplZML\\Lib\\site-packages\\pip\\_vendor\\rich\\_win32_console.py\", line 403, in write_text\n",
      "    self.write(text)\n",
      "  File \"C:\\Users\\quocd\\AppData\\Roaming\\uv\\python\\cpython-3.12.8-windows-x86_64-none\\Lib\\encodings\\cp1252.py\", line 19, in encode\n",
      "    return codecs.charmap_encode(input,self.errors,encoding_table)[0]\n",
      "           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "UnicodeEncodeError: 'charmap' codec can't encode character '\\U0001f680' in position 26: character maps to <undefined>\n",
      "Call stack:\n",
      "  File \"<frozen runpy>\", line 198, in _run_module_as_main\n",
      "  File \"<frozen runpy>\", line 88, in _run_code\n",
      "  File \"C:\\Users\\quocd\\AppData\\Local\\uv\\cache\\archive-v0\\z8lkNhNwG6XGWGBlplZML\\Scripts\\pip.exe\\__main__.py\", line 8, in <module>\n",
      "    sys.exit(main())\n",
      "  File \"C:\\Users\\quocd\\AppData\\Local\\uv\\cache\\archive-v0\\z8lkNhNwG6XGWGBlplZML\\Lib\\site-packages\\pip\\_internal\\cli\\main.py\", line 80, in main\n",
      "    return command.main(cmd_args)\n",
      "  File \"C:\\Users\\quocd\\AppData\\Local\\uv\\cache\\archive-v0\\z8lkNhNwG6XGWGBlplZML\\Lib\\site-packages\\pip\\_internal\\cli\\base_command.py\", line 157, in main\n",
      "    return self._main(args)\n",
      "  File \"C:\\Users\\quocd\\AppData\\Local\\uv\\cache\\archive-v0\\z8lkNhNwG6XGWGBlplZML\\Lib\\site-packages\\pip\\_internal\\cli\\base_command.py\", line 231, in _main\n",
      "    return self._run_wrapper(level_number, options, args)\n",
      "  File \"C:\\Users\\quocd\\AppData\\Local\\uv\\cache\\archive-v0\\z8lkNhNwG6XGWGBlplZML\\Lib\\site-packages\\pip\\_internal\\cli\\base_command.py\", line 105, in _run_wrapper\n",
      "    status = _inner_run()\n",
      "  File \"C:\\Users\\quocd\\AppData\\Local\\uv\\cache\\archive-v0\\z8lkNhNwG6XGWGBlplZML\\Lib\\site-packages\\pip\\_internal\\cli\\base_command.py\", line 96, in _inner_run\n",
      "    return self.run(options, args)\n",
      "  File \"C:\\Users\\quocd\\AppData\\Local\\uv\\cache\\archive-v0\\z8lkNhNwG6XGWGBlplZML\\Lib\\site-packages\\pip\\_internal\\commands\\show.py\", line 46, in run\n",
      "    if not print_results(\n",
      "  File \"C:\\Users\\quocd\\AppData\\Local\\uv\\cache\\archive-v0\\z8lkNhNwG6XGWGBlplZML\\Lib\\site-packages\\pip\\_internal\\commands\\show.py\", line 189, in print_results\n",
      "    write_output(\"License: %s\", dist.license)\n",
      "  File \"C:\\Users\\quocd\\AppData\\Local\\uv\\cache\\archive-v0\\z8lkNhNwG6XGWGBlplZML\\Lib\\site-packages\\pip\\_internal\\utils\\misc.py\", line 374, in write_output\n",
      "    logger.info(msg, *args)\n",
      "  File \"C:\\Users\\quocd\\AppData\\Roaming\\uv\\python\\cpython-3.12.8-windows-x86_64-none\\Lib\\logging\\__init__.py\", line 1539, in info\n",
      "    self._log(INFO, msg, args, **kwargs)\n",
      "  File \"C:\\Users\\quocd\\AppData\\Roaming\\uv\\python\\cpython-3.12.8-windows-x86_64-none\\Lib\\logging\\__init__.py\", line 1684, in _log\n",
      "    self.handle(record)\n",
      "  File \"C:\\Users\\quocd\\AppData\\Roaming\\uv\\python\\cpython-3.12.8-windows-x86_64-none\\Lib\\logging\\__init__.py\", line 1700, in handle\n",
      "    self.callHandlers(record)\n",
      "  File \"C:\\Users\\quocd\\AppData\\Roaming\\uv\\python\\cpython-3.12.8-windows-x86_64-none\\Lib\\logging\\__init__.py\", line 1762, in callHandlers\n",
      "    hdlr.handle(record)\n",
      "  File \"C:\\Users\\quocd\\AppData\\Roaming\\uv\\python\\cpython-3.12.8-windows-x86_64-none\\Lib\\logging\\__init__.py\", line 1028, in handle\n",
      "    self.emit(record)\n",
      "  File \"C:\\Users\\quocd\\AppData\\Local\\uv\\cache\\archive-v0\\z8lkNhNwG6XGWGBlplZML\\Lib\\site-packages\\pip\\_internal\\utils\\logging.py\", line 179, in emit\n",
      "    self.handleError(record)\n",
      "Message: 'License: %s'\n",
      "Arguments: ('AGPL-3.0',)\n",
      "--- Logging error ---\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\quocd\\AppData\\Local\\uv\\cache\\archive-v0\\z8lkNhNwG6XGWGBlplZML\\Lib\\site-packages\\pip\\_internal\\utils\\logging.py\", line 177, in emit\n",
      "    self.console.print(renderable, overflow=\"ignore\", crop=False, style=style)\n",
      "  File \"C:\\Users\\quocd\\AppData\\Local\\uv\\cache\\archive-v0\\z8lkNhNwG6XGWGBlplZML\\Lib\\site-packages\\pip\\_vendor\\rich\\console.py\", line 1673, in print\n",
      "    with self:\n",
      "         ^^^^\n",
      "  File \"C:\\Users\\quocd\\AppData\\Local\\uv\\cache\\archive-v0\\z8lkNhNwG6XGWGBlplZML\\Lib\\site-packages\\pip\\_vendor\\rich\\console.py\", line 865, in __exit__\n",
      "    self._exit_buffer()\n",
      "  File \"C:\\Users\\quocd\\AppData\\Local\\uv\\cache\\archive-v0\\z8lkNhNwG6XGWGBlplZML\\Lib\\site-packages\\pip\\_vendor\\rich\\console.py\", line 823, in _exit_buffer\n",
      "    self._check_buffer()\n",
      "  File \"C:\\Users\\quocd\\AppData\\Local\\uv\\cache\\archive-v0\\z8lkNhNwG6XGWGBlplZML\\Lib\\site-packages\\pip\\_vendor\\rich\\console.py\", line 2027, in _check_buffer\n",
      "    legacy_windows_render(buffer, LegacyWindowsTerm(self.file))\n",
      "  File \"C:\\Users\\quocd\\AppData\\Local\\uv\\cache\\archive-v0\\z8lkNhNwG6XGWGBlplZML\\Lib\\site-packages\\pip\\_vendor\\rich\\_windows_renderer.py\", line 19, in legacy_windows_render\n",
      "    term.write_text(text)\n",
      "  File \"C:\\Users\\quocd\\AppData\\Local\\uv\\cache\\archive-v0\\z8lkNhNwG6XGWGBlplZML\\Lib\\site-packages\\pip\\_vendor\\rich\\_win32_console.py\", line 403, in write_text\n",
      "    self.write(text)\n",
      "  File \"C:\\Users\\quocd\\AppData\\Roaming\\uv\\python\\cpython-3.12.8-windows-x86_64-none\\Lib\\encodings\\cp1252.py\", line 19, in encode\n",
      "    return codecs.charmap_encode(input,self.errors,encoding_table)[0]\n",
      "           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "UnicodeEncodeError: 'charmap' codec can't encode character '\\U0001f680' in position 26: character maps to <undefined>\n",
      "Call stack:\n",
      "  File \"<frozen runpy>\", line 198, in _run_module_as_main\n",
      "  File \"<frozen runpy>\", line 88, in _run_code\n",
      "  File \"C:\\Users\\quocd\\AppData\\Local\\uv\\cache\\archive-v0\\z8lkNhNwG6XGWGBlplZML\\Scripts\\pip.exe\\__main__.py\", line 8, in <module>\n",
      "    sys.exit(main())\n",
      "  File \"C:\\Users\\quocd\\AppData\\Local\\uv\\cache\\archive-v0\\z8lkNhNwG6XGWGBlplZML\\Lib\\site-packages\\pip\\_internal\\cli\\main.py\", line 80, in main\n",
      "    return command.main(cmd_args)\n",
      "  File \"C:\\Users\\quocd\\AppData\\Local\\uv\\cache\\archive-v0\\z8lkNhNwG6XGWGBlplZML\\Lib\\site-packages\\pip\\_internal\\cli\\base_command.py\", line 157, in main\n",
      "    return self._main(args)\n",
      "  File \"C:\\Users\\quocd\\AppData\\Local\\uv\\cache\\archive-v0\\z8lkNhNwG6XGWGBlplZML\\Lib\\site-packages\\pip\\_internal\\cli\\base_command.py\", line 231, in _main\n",
      "    return self._run_wrapper(level_number, options, args)\n",
      "  File \"C:\\Users\\quocd\\AppData\\Local\\uv\\cache\\archive-v0\\z8lkNhNwG6XGWGBlplZML\\Lib\\site-packages\\pip\\_internal\\cli\\base_command.py\", line 105, in _run_wrapper\n",
      "    status = _inner_run()\n",
      "  File \"C:\\Users\\quocd\\AppData\\Local\\uv\\cache\\archive-v0\\z8lkNhNwG6XGWGBlplZML\\Lib\\site-packages\\pip\\_internal\\cli\\base_command.py\", line 96, in _inner_run\n",
      "    return self.run(options, args)\n",
      "  File \"C:\\Users\\quocd\\AppData\\Local\\uv\\cache\\archive-v0\\z8lkNhNwG6XGWGBlplZML\\Lib\\site-packages\\pip\\_internal\\commands\\show.py\", line 46, in run\n",
      "    if not print_results(\n",
      "  File \"C:\\Users\\quocd\\AppData\\Local\\uv\\cache\\archive-v0\\z8lkNhNwG6XGWGBlplZML\\Lib\\site-packages\\pip\\_internal\\commands\\show.py\", line 190, in print_results\n",
      "    write_output(\"Location: %s\", dist.location)\n",
      "  File \"C:\\Users\\quocd\\AppData\\Local\\uv\\cache\\archive-v0\\z8lkNhNwG6XGWGBlplZML\\Lib\\site-packages\\pip\\_internal\\utils\\misc.py\", line 374, in write_output\n",
      "    logger.info(msg, *args)\n",
      "  File \"C:\\Users\\quocd\\AppData\\Roaming\\uv\\python\\cpython-3.12.8-windows-x86_64-none\\Lib\\logging\\__init__.py\", line 1539, in info\n",
      "    self._log(INFO, msg, args, **kwargs)\n",
      "  File \"C:\\Users\\quocd\\AppData\\Roaming\\uv\\python\\cpython-3.12.8-windows-x86_64-none\\Lib\\logging\\__init__.py\", line 1684, in _log\n",
      "    self.handle(record)\n",
      "  File \"C:\\Users\\quocd\\AppData\\Roaming\\uv\\python\\cpython-3.12.8-windows-x86_64-none\\Lib\\logging\\__init__.py\", line 1700, in handle\n",
      "    self.callHandlers(record)\n",
      "  File \"C:\\Users\\quocd\\AppData\\Roaming\\uv\\python\\cpython-3.12.8-windows-x86_64-none\\Lib\\logging\\__init__.py\", line 1762, in callHandlers\n",
      "    hdlr.handle(record)\n",
      "  File \"C:\\Users\\quocd\\AppData\\Roaming\\uv\\python\\cpython-3.12.8-windows-x86_64-none\\Lib\\logging\\__init__.py\", line 1028, in handle\n",
      "    self.emit(record)\n",
      "  File \"C:\\Users\\quocd\\AppData\\Local\\uv\\cache\\archive-v0\\z8lkNhNwG6XGWGBlplZML\\Lib\\site-packages\\pip\\_internal\\utils\\logging.py\", line 179, in emit\n",
      "    self.handleError(record)\n",
      "Message: 'Location: %s'\n",
      "Arguments: ('C:\\\\Users\\\\quocd\\\\AppData\\\\Local\\\\uv\\\\cache\\\\archive-v0\\\\z8lkNhNwG6XGWGBlplZML\\\\Lib\\\\site-packages',)\n",
      "--- Logging error ---\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\quocd\\AppData\\Local\\uv\\cache\\archive-v0\\z8lkNhNwG6XGWGBlplZML\\Lib\\site-packages\\pip\\_internal\\utils\\logging.py\", line 177, in emit\n",
      "    self.console.print(renderable, overflow=\"ignore\", crop=False, style=style)\n",
      "  File \"C:\\Users\\quocd\\AppData\\Local\\uv\\cache\\archive-v0\\z8lkNhNwG6XGWGBlplZML\\Lib\\site-packages\\pip\\_vendor\\rich\\console.py\", line 1673, in print\n",
      "    with self:\n",
      "         ^^^^\n",
      "  File \"C:\\Users\\quocd\\AppData\\Local\\uv\\cache\\archive-v0\\z8lkNhNwG6XGWGBlplZML\\Lib\\site-packages\\pip\\_vendor\\rich\\console.py\", line 865, in __exit__\n",
      "    self._exit_buffer()\n",
      "  File \"C:\\Users\\quocd\\AppData\\Local\\uv\\cache\\archive-v0\\z8lkNhNwG6XGWGBlplZML\\Lib\\site-packages\\pip\\_vendor\\rich\\console.py\", line 823, in _exit_buffer\n",
      "    self._check_buffer()\n",
      "  File \"C:\\Users\\quocd\\AppData\\Local\\uv\\cache\\archive-v0\\z8lkNhNwG6XGWGBlplZML\\Lib\\site-packages\\pip\\_vendor\\rich\\console.py\", line 2027, in _check_buffer\n",
      "    legacy_windows_render(buffer, LegacyWindowsTerm(self.file))\n",
      "  File \"C:\\Users\\quocd\\AppData\\Local\\uv\\cache\\archive-v0\\z8lkNhNwG6XGWGBlplZML\\Lib\\site-packages\\pip\\_vendor\\rich\\_windows_renderer.py\", line 19, in legacy_windows_render\n",
      "    term.write_text(text)\n",
      "  File \"C:\\Users\\quocd\\AppData\\Local\\uv\\cache\\archive-v0\\z8lkNhNwG6XGWGBlplZML\\Lib\\site-packages\\pip\\_vendor\\rich\\_win32_console.py\", line 403, in write_text\n",
      "    self.write(text)\n",
      "  File \"C:\\Users\\quocd\\AppData\\Roaming\\uv\\python\\cpython-3.12.8-windows-x86_64-none\\Lib\\encodings\\cp1252.py\", line 19, in encode\n",
      "    return codecs.charmap_encode(input,self.errors,encoding_table)[0]\n",
      "           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "UnicodeEncodeError: 'charmap' codec can't encode character '\\U0001f680' in position 26: character maps to <undefined>\n",
      "Call stack:\n",
      "  File \"<frozen runpy>\", line 198, in _run_module_as_main\n",
      "  File \"<frozen runpy>\", line 88, in _run_code\n",
      "  File \"C:\\Users\\quocd\\AppData\\Local\\uv\\cache\\archive-v0\\z8lkNhNwG6XGWGBlplZML\\Scripts\\pip.exe\\__main__.py\", line 8, in <module>\n",
      "    sys.exit(main())\n",
      "  File \"C:\\Users\\quocd\\AppData\\Local\\uv\\cache\\archive-v0\\z8lkNhNwG6XGWGBlplZML\\Lib\\site-packages\\pip\\_internal\\cli\\main.py\", line 80, in main\n",
      "    return command.main(cmd_args)\n",
      "  File \"C:\\Users\\quocd\\AppData\\Local\\uv\\cache\\archive-v0\\z8lkNhNwG6XGWGBlplZML\\Lib\\site-packages\\pip\\_internal\\cli\\base_command.py\", line 157, in main\n",
      "    return self._main(args)\n",
      "  File \"C:\\Users\\quocd\\AppData\\Local\\uv\\cache\\archive-v0\\z8lkNhNwG6XGWGBlplZML\\Lib\\site-packages\\pip\\_internal\\cli\\base_command.py\", line 231, in _main\n",
      "    return self._run_wrapper(level_number, options, args)\n",
      "  File \"C:\\Users\\quocd\\AppData\\Local\\uv\\cache\\archive-v0\\z8lkNhNwG6XGWGBlplZML\\Lib\\site-packages\\pip\\_internal\\cli\\base_command.py\", line 105, in _run_wrapper\n",
      "    status = _inner_run()\n",
      "  File \"C:\\Users\\quocd\\AppData\\Local\\uv\\cache\\archive-v0\\z8lkNhNwG6XGWGBlplZML\\Lib\\site-packages\\pip\\_internal\\cli\\base_command.py\", line 96, in _inner_run\n",
      "    return self.run(options, args)\n",
      "  File \"C:\\Users\\quocd\\AppData\\Local\\uv\\cache\\archive-v0\\z8lkNhNwG6XGWGBlplZML\\Lib\\site-packages\\pip\\_internal\\commands\\show.py\", line 46, in run\n",
      "    if not print_results(\n",
      "  File \"C:\\Users\\quocd\\AppData\\Local\\uv\\cache\\archive-v0\\z8lkNhNwG6XGWGBlplZML\\Lib\\site-packages\\pip\\_internal\\commands\\show.py\", line 195, in print_results\n",
      "    write_output(\"Requires: %s\", \", \".join(dist.requires))\n",
      "  File \"C:\\Users\\quocd\\AppData\\Local\\uv\\cache\\archive-v0\\z8lkNhNwG6XGWGBlplZML\\Lib\\site-packages\\pip\\_internal\\utils\\misc.py\", line 374, in write_output\n",
      "    logger.info(msg, *args)\n",
      "  File \"C:\\Users\\quocd\\AppData\\Roaming\\uv\\python\\cpython-3.12.8-windows-x86_64-none\\Lib\\logging\\__init__.py\", line 1539, in info\n",
      "    self._log(INFO, msg, args, **kwargs)\n",
      "  File \"C:\\Users\\quocd\\AppData\\Roaming\\uv\\python\\cpython-3.12.8-windows-x86_64-none\\Lib\\logging\\__init__.py\", line 1684, in _log\n",
      "    self.handle(record)\n",
      "  File \"C:\\Users\\quocd\\AppData\\Roaming\\uv\\python\\cpython-3.12.8-windows-x86_64-none\\Lib\\logging\\__init__.py\", line 1700, in handle\n",
      "    self.callHandlers(record)\n",
      "  File \"C:\\Users\\quocd\\AppData\\Roaming\\uv\\python\\cpython-3.12.8-windows-x86_64-none\\Lib\\logging\\__init__.py\", line 1762, in callHandlers\n",
      "    hdlr.handle(record)\n",
      "  File \"C:\\Users\\quocd\\AppData\\Roaming\\uv\\python\\cpython-3.12.8-windows-x86_64-none\\Lib\\logging\\__init__.py\", line 1028, in handle\n",
      "    self.emit(record)\n",
      "  File \"C:\\Users\\quocd\\AppData\\Local\\uv\\cache\\archive-v0\\z8lkNhNwG6XGWGBlplZML\\Lib\\site-packages\\pip\\_internal\\utils\\logging.py\", line 179, in emit\n",
      "    self.handleError(record)\n",
      "Message: 'Requires: %s'\n",
      "Arguments: ('matplotlib, numpy, opencv-python, pandas, pillow, psutil, py-cpuinfo, pyyaml, requests, scipy, seaborn, torch, torchvision, tqdm, ultralytics-thop',)\n",
      "--- Logging error ---\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\quocd\\AppData\\Local\\uv\\cache\\archive-v0\\z8lkNhNwG6XGWGBlplZML\\Lib\\site-packages\\pip\\_internal\\utils\\logging.py\", line 177, in emit\n",
      "    self.console.print(renderable, overflow=\"ignore\", crop=False, style=style)\n",
      "  File \"C:\\Users\\quocd\\AppData\\Local\\uv\\cache\\archive-v0\\z8lkNhNwG6XGWGBlplZML\\Lib\\site-packages\\pip\\_vendor\\rich\\console.py\", line 1673, in print\n",
      "    with self:\n",
      "         ^^^^\n",
      "  File \"C:\\Users\\quocd\\AppData\\Local\\uv\\cache\\archive-v0\\z8lkNhNwG6XGWGBlplZML\\Lib\\site-packages\\pip\\_vendor\\rich\\console.py\", line 865, in __exit__\n",
      "    self._exit_buffer()\n",
      "  File \"C:\\Users\\quocd\\AppData\\Local\\uv\\cache\\archive-v0\\z8lkNhNwG6XGWGBlplZML\\Lib\\site-packages\\pip\\_vendor\\rich\\console.py\", line 823, in _exit_buffer\n",
      "    self._check_buffer()\n",
      "  File \"C:\\Users\\quocd\\AppData\\Local\\uv\\cache\\archive-v0\\z8lkNhNwG6XGWGBlplZML\\Lib\\site-packages\\pip\\_vendor\\rich\\console.py\", line 2027, in _check_buffer\n",
      "    legacy_windows_render(buffer, LegacyWindowsTerm(self.file))\n",
      "  File \"C:\\Users\\quocd\\AppData\\Local\\uv\\cache\\archive-v0\\z8lkNhNwG6XGWGBlplZML\\Lib\\site-packages\\pip\\_vendor\\rich\\_windows_renderer.py\", line 19, in legacy_windows_render\n",
      "    term.write_text(text)\n",
      "  File \"C:\\Users\\quocd\\AppData\\Local\\uv\\cache\\archive-v0\\z8lkNhNwG6XGWGBlplZML\\Lib\\site-packages\\pip\\_vendor\\rich\\_win32_console.py\", line 403, in write_text\n",
      "    self.write(text)\n",
      "  File \"C:\\Users\\quocd\\AppData\\Roaming\\uv\\python\\cpython-3.12.8-windows-x86_64-none\\Lib\\encodings\\cp1252.py\", line 19, in encode\n",
      "    return codecs.charmap_encode(input,self.errors,encoding_table)[0]\n",
      "           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "UnicodeEncodeError: 'charmap' codec can't encode character '\\U0001f680' in position 26: character maps to <undefined>\n",
      "Call stack:\n",
      "  File \"<frozen runpy>\", line 198, in _run_module_as_main\n",
      "  File \"<frozen runpy>\", line 88, in _run_code\n",
      "  File \"C:\\Users\\quocd\\AppData\\Local\\uv\\cache\\archive-v0\\z8lkNhNwG6XGWGBlplZML\\Scripts\\pip.exe\\__main__.py\", line 8, in <module>\n",
      "    sys.exit(main())\n",
      "  File \"C:\\Users\\quocd\\AppData\\Local\\uv\\cache\\archive-v0\\z8lkNhNwG6XGWGBlplZML\\Lib\\site-packages\\pip\\_internal\\cli\\main.py\", line 80, in main\n",
      "    return command.main(cmd_args)\n",
      "  File \"C:\\Users\\quocd\\AppData\\Local\\uv\\cache\\archive-v0\\z8lkNhNwG6XGWGBlplZML\\Lib\\site-packages\\pip\\_internal\\cli\\base_command.py\", line 157, in main\n",
      "    return self._main(args)\n",
      "  File \"C:\\Users\\quocd\\AppData\\Local\\uv\\cache\\archive-v0\\z8lkNhNwG6XGWGBlplZML\\Lib\\site-packages\\pip\\_internal\\cli\\base_command.py\", line 231, in _main\n",
      "    return self._run_wrapper(level_number, options, args)\n",
      "  File \"C:\\Users\\quocd\\AppData\\Local\\uv\\cache\\archive-v0\\z8lkNhNwG6XGWGBlplZML\\Lib\\site-packages\\pip\\_internal\\cli\\base_command.py\", line 105, in _run_wrapper\n",
      "    status = _inner_run()\n",
      "  File \"C:\\Users\\quocd\\AppData\\Local\\uv\\cache\\archive-v0\\z8lkNhNwG6XGWGBlplZML\\Lib\\site-packages\\pip\\_internal\\cli\\base_command.py\", line 96, in _inner_run\n",
      "    return self.run(options, args)\n",
      "  File \"C:\\Users\\quocd\\AppData\\Local\\uv\\cache\\archive-v0\\z8lkNhNwG6XGWGBlplZML\\Lib\\site-packages\\pip\\_internal\\commands\\show.py\", line 46, in run\n",
      "    if not print_results(\n",
      "  File \"C:\\Users\\quocd\\AppData\\Local\\uv\\cache\\archive-v0\\z8lkNhNwG6XGWGBlplZML\\Lib\\site-packages\\pip\\_internal\\commands\\show.py\", line 196, in print_results\n",
      "    write_output(\"Required-by: %s\", \", \".join(dist.required_by))\n",
      "  File \"C:\\Users\\quocd\\AppData\\Local\\uv\\cache\\archive-v0\\z8lkNhNwG6XGWGBlplZML\\Lib\\site-packages\\pip\\_internal\\utils\\misc.py\", line 374, in write_output\n",
      "    logger.info(msg, *args)\n",
      "  File \"C:\\Users\\quocd\\AppData\\Roaming\\uv\\python\\cpython-3.12.8-windows-x86_64-none\\Lib\\logging\\__init__.py\", line 1539, in info\n",
      "    self._log(INFO, msg, args, **kwargs)\n",
      "  File \"C:\\Users\\quocd\\AppData\\Roaming\\uv\\python\\cpython-3.12.8-windows-x86_64-none\\Lib\\logging\\__init__.py\", line 1684, in _log\n",
      "    self.handle(record)\n",
      "  File \"C:\\Users\\quocd\\AppData\\Roaming\\uv\\python\\cpython-3.12.8-windows-x86_64-none\\Lib\\logging\\__init__.py\", line 1700, in handle\n",
      "    self.callHandlers(record)\n",
      "  File \"C:\\Users\\quocd\\AppData\\Roaming\\uv\\python\\cpython-3.12.8-windows-x86_64-none\\Lib\\logging\\__init__.py\", line 1762, in callHandlers\n",
      "    hdlr.handle(record)\n",
      "  File \"C:\\Users\\quocd\\AppData\\Roaming\\uv\\python\\cpython-3.12.8-windows-x86_64-none\\Lib\\logging\\__init__.py\", line 1028, in handle\n",
      "    self.emit(record)\n",
      "  File \"C:\\Users\\quocd\\AppData\\Local\\uv\\cache\\archive-v0\\z8lkNhNwG6XGWGBlplZML\\Lib\\site-packages\\pip\\_internal\\utils\\logging.py\", line 179, in emit\n",
      "    self.handleError(record)\n",
      "Message: 'Required-by: %s'\n",
      "Arguments: ('',)\n"
     ]
    }
   ],
   "source": [
    "!pip show ultralytics "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CUDA is available: False\n"
     ]
    }
   ],
   "source": [
    "import torch\n",
    "\n",
    "# Check CUDA availability\n",
    "print(f\"CUDA is available: {torch.cuda.is_available()}\")\n",
    "\n",
    "# If you want more details:\n",
    "if torch.cuda.is_available():\n",
    "    print(f\"Current CUDA device: {torch.cuda.current_device()}\")\n",
    "    print(f\"Device name: {torch.cuda.get_device_name()}\")\n",
    "    print(f\"Number of CUDA devices: {torch.cuda.device_count()}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "[2025-01-14 16:48:45,907] [ WARNING] easyocr.py:80 - Neither CUDA nor MPS are available - defaulting to CPU. Note: This module is much faster with a GPU.\n"
     ]
    },
    {
     "ename": "AssertionError",
     "evalue": "Torch not compiled with CUDA enabled",
     "output_type": "error",
     "traceback": [
      "\u001b[1;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[1;31mAssertionError\u001b[0m                            Traceback (most recent call last)",
      "Cell \u001b[1;32mIn[2], line 10\u001b[0m\n\u001b[0;32m      6\u001b[0m model_path\u001b[38;5;241m=\u001b[39m\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mweights/icon_detect_v1_5/model_v1_5.pt\u001b[39m\u001b[38;5;124m'\u001b[39m\n\u001b[0;32m      8\u001b[0m som_model \u001b[38;5;241m=\u001b[39m get_yolo_model(model_path)\n\u001b[1;32m---> 10\u001b[0m \u001b[43msom_model\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mto\u001b[49m\u001b[43m(\u001b[49m\u001b[43mdevice\u001b[49m\u001b[43m)\u001b[49m\n\u001b[0;32m     11\u001b[0m \u001b[38;5;28mprint\u001b[39m(\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mmodel to \u001b[39m\u001b[38;5;132;01m{}\u001b[39;00m\u001b[38;5;124m'\u001b[39m\u001b[38;5;241m.\u001b[39mformat(device))\n",
      "File \u001b[1;32m~\\AppData\\Local\\uv\\cache\\archive-v0\\z8lkNhNwG6XGWGBlplZML\\Lib\\site-packages\\torch\\nn\\modules\\module.py:1340\u001b[0m, in \u001b[0;36mModule.to\u001b[1;34m(self, *args, **kwargs)\u001b[0m\n\u001b[0;32m   1337\u001b[0m         \u001b[38;5;28;01melse\u001b[39;00m:\n\u001b[0;32m   1338\u001b[0m             \u001b[38;5;28;01mraise\u001b[39;00m\n\u001b[1;32m-> 1340\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_apply\u001b[49m\u001b[43m(\u001b[49m\u001b[43mconvert\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[1;32m~\\AppData\\Local\\uv\\cache\\archive-v0\\z8lkNhNwG6XGWGBlplZML\\Lib\\site-packages\\ultralytics\\engine\\model.py:882\u001b[0m, in \u001b[0;36mModel._apply\u001b[1;34m(self, fn)\u001b[0m\n\u001b[0;32m    860\u001b[0m \u001b[38;5;250m\u001b[39m\u001b[38;5;124;03m\"\"\"\u001b[39;00m\n\u001b[0;32m    861\u001b[0m \u001b[38;5;124;03mApplies a function to model tensors that are not parameters or registered buffers.\u001b[39;00m\n\u001b[0;32m    862\u001b[0m \n\u001b[1;32m   (...)\u001b[0m\n\u001b[0;32m    879\u001b[0m \u001b[38;5;124;03m    >>> model = model._apply(lambda t: t.cuda())  # Move model to GPU\u001b[39;00m\n\u001b[0;32m    880\u001b[0m \u001b[38;5;124;03m\"\"\"\u001b[39;00m\n\u001b[0;32m    881\u001b[0m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_check_is_pytorch_model()\n\u001b[1;32m--> 882\u001b[0m \u001b[38;5;28mself\u001b[39m \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;43msuper\u001b[39;49m\u001b[43m(\u001b[49m\u001b[43m)\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_apply\u001b[49m\u001b[43m(\u001b[49m\u001b[43mfn\u001b[49m\u001b[43m)\u001b[49m  \u001b[38;5;66;03m# noqa\u001b[39;00m\n\u001b[0;32m    883\u001b[0m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mpredictor \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;01mNone\u001b[39;00m  \u001b[38;5;66;03m# reset predictor as device may have changed\u001b[39;00m\n\u001b[0;32m    884\u001b[0m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39moverrides[\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mdevice\u001b[39m\u001b[38;5;124m\"\u001b[39m] \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mdevice  \u001b[38;5;66;03m# was str(self.device) i.e. device(type='cuda', index=0) -> 'cuda:0'\u001b[39;00m\n",
      "File \u001b[1;32m~\\AppData\\Local\\uv\\cache\\archive-v0\\z8lkNhNwG6XGWGBlplZML\\Lib\\site-packages\\torch\\nn\\modules\\module.py:900\u001b[0m, in \u001b[0;36mModule._apply\u001b[1;34m(self, fn, recurse)\u001b[0m\n\u001b[0;32m    898\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m recurse:\n\u001b[0;32m    899\u001b[0m     \u001b[38;5;28;01mfor\u001b[39;00m module \u001b[38;5;129;01min\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mchildren():\n\u001b[1;32m--> 900\u001b[0m         \u001b[43mmodule\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_apply\u001b[49m\u001b[43m(\u001b[49m\u001b[43mfn\u001b[49m\u001b[43m)\u001b[49m\n\u001b[0;32m    902\u001b[0m \u001b[38;5;28;01mdef\u001b[39;00m\u001b[38;5;250m \u001b[39m\u001b[38;5;21mcompute_should_use_set_data\u001b[39m(tensor, tensor_applied):\n\u001b[0;32m    903\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m torch\u001b[38;5;241m.\u001b[39m_has_compatible_shallow_copy_type(tensor, tensor_applied):\n\u001b[0;32m    904\u001b[0m         \u001b[38;5;66;03m# If the new tensor has compatible tensor type as the existing tensor,\u001b[39;00m\n\u001b[0;32m    905\u001b[0m         \u001b[38;5;66;03m# the current behavior is to change the tensor in-place using `.data =`,\u001b[39;00m\n\u001b[1;32m   (...)\u001b[0m\n\u001b[0;32m    910\u001b[0m         \u001b[38;5;66;03m# global flag to let the user control whether they want the future\u001b[39;00m\n\u001b[0;32m    911\u001b[0m         \u001b[38;5;66;03m# behavior of overwriting the existing tensor or not.\u001b[39;00m\n",
      "File \u001b[1;32m~\\AppData\\Local\\uv\\cache\\archive-v0\\z8lkNhNwG6XGWGBlplZML\\Lib\\site-packages\\ultralytics\\nn\\tasks.py:260\u001b[0m, in \u001b[0;36mBaseModel._apply\u001b[1;34m(self, fn)\u001b[0m\n\u001b[0;32m    250\u001b[0m \u001b[38;5;28;01mdef\u001b[39;00m\u001b[38;5;250m \u001b[39m\u001b[38;5;21m_apply\u001b[39m(\u001b[38;5;28mself\u001b[39m, fn):\n\u001b[0;32m    251\u001b[0m \u001b[38;5;250m    \u001b[39m\u001b[38;5;124;03m\"\"\"\u001b[39;00m\n\u001b[0;32m    252\u001b[0m \u001b[38;5;124;03m    Applies a function to all the tensors in the model that are not parameters or registered buffers.\u001b[39;00m\n\u001b[0;32m    253\u001b[0m \n\u001b[1;32m   (...)\u001b[0m\n\u001b[0;32m    258\u001b[0m \u001b[38;5;124;03m        (BaseModel): An updated BaseModel object.\u001b[39;00m\n\u001b[0;32m    259\u001b[0m \u001b[38;5;124;03m    \"\"\"\u001b[39;00m\n\u001b[1;32m--> 260\u001b[0m     \u001b[38;5;28mself\u001b[39m \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;43msuper\u001b[39;49m\u001b[43m(\u001b[49m\u001b[43m)\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_apply\u001b[49m\u001b[43m(\u001b[49m\u001b[43mfn\u001b[49m\u001b[43m)\u001b[49m\n\u001b[0;32m    261\u001b[0m     m \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mmodel[\u001b[38;5;241m-\u001b[39m\u001b[38;5;241m1\u001b[39m]  \u001b[38;5;66;03m# Detect()\u001b[39;00m\n\u001b[0;32m    262\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;28misinstance\u001b[39m(m, Detect):  \u001b[38;5;66;03m# includes all Detect subclasses like Segment, Pose, OBB, WorldDetect\u001b[39;00m\n",
      "File \u001b[1;32m~\\AppData\\Local\\uv\\cache\\archive-v0\\z8lkNhNwG6XGWGBlplZML\\Lib\\site-packages\\torch\\nn\\modules\\module.py:900\u001b[0m, in \u001b[0;36mModule._apply\u001b[1;34m(self, fn, recurse)\u001b[0m\n\u001b[0;32m    898\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m recurse:\n\u001b[0;32m    899\u001b[0m     \u001b[38;5;28;01mfor\u001b[39;00m module \u001b[38;5;129;01min\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mchildren():\n\u001b[1;32m--> 900\u001b[0m         \u001b[43mmodule\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_apply\u001b[49m\u001b[43m(\u001b[49m\u001b[43mfn\u001b[49m\u001b[43m)\u001b[49m\n\u001b[0;32m    902\u001b[0m \u001b[38;5;28;01mdef\u001b[39;00m\u001b[38;5;250m \u001b[39m\u001b[38;5;21mcompute_should_use_set_data\u001b[39m(tensor, tensor_applied):\n\u001b[0;32m    903\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m torch\u001b[38;5;241m.\u001b[39m_has_compatible_shallow_copy_type(tensor, tensor_applied):\n\u001b[0;32m    904\u001b[0m         \u001b[38;5;66;03m# If the new tensor has compatible tensor type as the existing tensor,\u001b[39;00m\n\u001b[0;32m    905\u001b[0m         \u001b[38;5;66;03m# the current behavior is to change the tensor in-place using `.data =`,\u001b[39;00m\n\u001b[1;32m   (...)\u001b[0m\n\u001b[0;32m    910\u001b[0m         \u001b[38;5;66;03m# global flag to let the user control whether they want the future\u001b[39;00m\n\u001b[0;32m    911\u001b[0m         \u001b[38;5;66;03m# behavior of overwriting the existing tensor or not.\u001b[39;00m\n",
      "File \u001b[1;32m~\\AppData\\Local\\uv\\cache\\archive-v0\\z8lkNhNwG6XGWGBlplZML\\Lib\\site-packages\\torch\\nn\\modules\\module.py:900\u001b[0m, in \u001b[0;36mModule._apply\u001b[1;34m(self, fn, recurse)\u001b[0m\n\u001b[0;32m    898\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m recurse:\n\u001b[0;32m    899\u001b[0m     \u001b[38;5;28;01mfor\u001b[39;00m module \u001b[38;5;129;01min\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mchildren():\n\u001b[1;32m--> 900\u001b[0m         \u001b[43mmodule\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_apply\u001b[49m\u001b[43m(\u001b[49m\u001b[43mfn\u001b[49m\u001b[43m)\u001b[49m\n\u001b[0;32m    902\u001b[0m \u001b[38;5;28;01mdef\u001b[39;00m\u001b[38;5;250m \u001b[39m\u001b[38;5;21mcompute_should_use_set_data\u001b[39m(tensor, tensor_applied):\n\u001b[0;32m    903\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m torch\u001b[38;5;241m.\u001b[39m_has_compatible_shallow_copy_type(tensor, tensor_applied):\n\u001b[0;32m    904\u001b[0m         \u001b[38;5;66;03m# If the new tensor has compatible tensor type as the existing tensor,\u001b[39;00m\n\u001b[0;32m    905\u001b[0m         \u001b[38;5;66;03m# the current behavior is to change the tensor in-place using `.data =`,\u001b[39;00m\n\u001b[1;32m   (...)\u001b[0m\n\u001b[0;32m    910\u001b[0m         \u001b[38;5;66;03m# global flag to let the user control whether they want the future\u001b[39;00m\n\u001b[0;32m    911\u001b[0m         \u001b[38;5;66;03m# behavior of overwriting the existing tensor or not.\u001b[39;00m\n",
      "File \u001b[1;32m~\\AppData\\Local\\uv\\cache\\archive-v0\\z8lkNhNwG6XGWGBlplZML\\Lib\\site-packages\\torch\\nn\\modules\\module.py:900\u001b[0m, in \u001b[0;36mModule._apply\u001b[1;34m(self, fn, recurse)\u001b[0m\n\u001b[0;32m    898\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m recurse:\n\u001b[0;32m    899\u001b[0m     \u001b[38;5;28;01mfor\u001b[39;00m module \u001b[38;5;129;01min\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mchildren():\n\u001b[1;32m--> 900\u001b[0m         \u001b[43mmodule\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_apply\u001b[49m\u001b[43m(\u001b[49m\u001b[43mfn\u001b[49m\u001b[43m)\u001b[49m\n\u001b[0;32m    902\u001b[0m \u001b[38;5;28;01mdef\u001b[39;00m\u001b[38;5;250m \u001b[39m\u001b[38;5;21mcompute_should_use_set_data\u001b[39m(tensor, tensor_applied):\n\u001b[0;32m    903\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m torch\u001b[38;5;241m.\u001b[39m_has_compatible_shallow_copy_type(tensor, tensor_applied):\n\u001b[0;32m    904\u001b[0m         \u001b[38;5;66;03m# If the new tensor has compatible tensor type as the existing tensor,\u001b[39;00m\n\u001b[0;32m    905\u001b[0m         \u001b[38;5;66;03m# the current behavior is to change the tensor in-place using `.data =`,\u001b[39;00m\n\u001b[1;32m   (...)\u001b[0m\n\u001b[0;32m    910\u001b[0m         \u001b[38;5;66;03m# global flag to let the user control whether they want the future\u001b[39;00m\n\u001b[0;32m    911\u001b[0m         \u001b[38;5;66;03m# behavior of overwriting the existing tensor or not.\u001b[39;00m\n",
      "File \u001b[1;32m~\\AppData\\Local\\uv\\cache\\archive-v0\\z8lkNhNwG6XGWGBlplZML\\Lib\\site-packages\\torch\\nn\\modules\\module.py:927\u001b[0m, in \u001b[0;36mModule._apply\u001b[1;34m(self, fn, recurse)\u001b[0m\n\u001b[0;32m    923\u001b[0m \u001b[38;5;66;03m# Tensors stored in modules are graph leaves, and we don't want to\u001b[39;00m\n\u001b[0;32m    924\u001b[0m \u001b[38;5;66;03m# track autograd history of `param_applied`, so we have to use\u001b[39;00m\n\u001b[0;32m    925\u001b[0m \u001b[38;5;66;03m# `with torch.no_grad():`\u001b[39;00m\n\u001b[0;32m    926\u001b[0m \u001b[38;5;28;01mwith\u001b[39;00m torch\u001b[38;5;241m.\u001b[39mno_grad():\n\u001b[1;32m--> 927\u001b[0m     param_applied \u001b[38;5;241m=\u001b[39m \u001b[43mfn\u001b[49m\u001b[43m(\u001b[49m\u001b[43mparam\u001b[49m\u001b[43m)\u001b[49m\n\u001b[0;32m    928\u001b[0m p_should_use_set_data \u001b[38;5;241m=\u001b[39m compute_should_use_set_data(param, param_applied)\n\u001b[0;32m    930\u001b[0m \u001b[38;5;66;03m# subclasses may have multiple child tensors so we need to use swap_tensors\u001b[39;00m\n",
      "File \u001b[1;32m~\\AppData\\Local\\uv\\cache\\archive-v0\\z8lkNhNwG6XGWGBlplZML\\Lib\\site-packages\\torch\\nn\\modules\\module.py:1326\u001b[0m, in \u001b[0;36mModule.to.<locals>.convert\u001b[1;34m(t)\u001b[0m\n\u001b[0;32m   1319\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m convert_to_format \u001b[38;5;129;01mis\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m \u001b[38;5;28;01mNone\u001b[39;00m \u001b[38;5;129;01mand\u001b[39;00m t\u001b[38;5;241m.\u001b[39mdim() \u001b[38;5;129;01min\u001b[39;00m (\u001b[38;5;241m4\u001b[39m, \u001b[38;5;241m5\u001b[39m):\n\u001b[0;32m   1320\u001b[0m         \u001b[38;5;28;01mreturn\u001b[39;00m t\u001b[38;5;241m.\u001b[39mto(\n\u001b[0;32m   1321\u001b[0m             device,\n\u001b[0;32m   1322\u001b[0m             dtype \u001b[38;5;28;01mif\u001b[39;00m t\u001b[38;5;241m.\u001b[39mis_floating_point() \u001b[38;5;129;01mor\u001b[39;00m t\u001b[38;5;241m.\u001b[39mis_complex() \u001b[38;5;28;01melse\u001b[39;00m \u001b[38;5;28;01mNone\u001b[39;00m,\n\u001b[0;32m   1323\u001b[0m             non_blocking,\n\u001b[0;32m   1324\u001b[0m             memory_format\u001b[38;5;241m=\u001b[39mconvert_to_format,\n\u001b[0;32m   1325\u001b[0m         )\n\u001b[1;32m-> 1326\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[43mt\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mto\u001b[49m\u001b[43m(\u001b[49m\n\u001b[0;32m   1327\u001b[0m \u001b[43m        \u001b[49m\u001b[43mdevice\u001b[49m\u001b[43m,\u001b[49m\n\u001b[0;32m   1328\u001b[0m \u001b[43m        \u001b[49m\u001b[43mdtype\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;28;43;01mif\u001b[39;49;00m\u001b[43m \u001b[49m\u001b[43mt\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mis_floating_point\u001b[49m\u001b[43m(\u001b[49m\u001b[43m)\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;129;43;01mor\u001b[39;49;00m\u001b[43m \u001b[49m\u001b[43mt\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mis_complex\u001b[49m\u001b[43m(\u001b[49m\u001b[43m)\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;28;43;01melse\u001b[39;49;00m\u001b[43m \u001b[49m\u001b[38;5;28;43;01mNone\u001b[39;49;00m\u001b[43m,\u001b[49m\n\u001b[0;32m   1329\u001b[0m \u001b[43m        \u001b[49m\u001b[43mnon_blocking\u001b[49m\u001b[43m,\u001b[49m\n\u001b[0;32m   1330\u001b[0m \u001b[43m    \u001b[49m\u001b[43m)\u001b[49m\n\u001b[0;32m   1331\u001b[0m \u001b[38;5;28;01mexcept\u001b[39;00m \u001b[38;5;167;01mNotImplementedError\u001b[39;00m \u001b[38;5;28;01mas\u001b[39;00m e:\n\u001b[0;32m   1332\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;28mstr\u001b[39m(e) \u001b[38;5;241m==\u001b[39m \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mCannot copy out of meta tensor; no data!\u001b[39m\u001b[38;5;124m\"\u001b[39m:\n",
      "File \u001b[1;32m~\\AppData\\Local\\uv\\cache\\archive-v0\\z8lkNhNwG6XGWGBlplZML\\Lib\\site-packages\\torch\\cuda\\__init__.py:310\u001b[0m, in \u001b[0;36m_lazy_init\u001b[1;34m()\u001b[0m\n\u001b[0;32m    305\u001b[0m     \u001b[38;5;28;01mraise\u001b[39;00m \u001b[38;5;167;01mRuntimeError\u001b[39;00m(\n\u001b[0;32m    306\u001b[0m         \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mCannot re-initialize CUDA in forked subprocess. To use CUDA with \u001b[39m\u001b[38;5;124m\"\u001b[39m\n\u001b[0;32m    307\u001b[0m         \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mmultiprocessing, you must use the \u001b[39m\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mspawn\u001b[39m\u001b[38;5;124m'\u001b[39m\u001b[38;5;124m start method\u001b[39m\u001b[38;5;124m\"\u001b[39m\n\u001b[0;32m    308\u001b[0m     )\n\u001b[0;32m    309\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m \u001b[38;5;28mhasattr\u001b[39m(torch\u001b[38;5;241m.\u001b[39m_C, \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124m_cuda_getDeviceCount\u001b[39m\u001b[38;5;124m\"\u001b[39m):\n\u001b[1;32m--> 310\u001b[0m     \u001b[38;5;28;01mraise\u001b[39;00m \u001b[38;5;167;01mAssertionError\u001b[39;00m(\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mTorch not compiled with CUDA enabled\u001b[39m\u001b[38;5;124m\"\u001b[39m)\n\u001b[0;32m    311\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m _cudart \u001b[38;5;129;01mis\u001b[39;00m \u001b[38;5;28;01mNone\u001b[39;00m:\n\u001b[0;32m    312\u001b[0m     \u001b[38;5;28;01mraise\u001b[39;00m \u001b[38;5;167;01mAssertionError\u001b[39;00m(\n\u001b[0;32m    313\u001b[0m         \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mlibcudart functions unavailable. It looks like you have a broken build?\u001b[39m\u001b[38;5;124m\"\u001b[39m\n\u001b[0;32m    314\u001b[0m     )\n",
      "\u001b[1;31mAssertionError\u001b[0m: Torch not compiled with CUDA enabled"
     ]
    }
   ],
   "source": [
    "from utils import get_som_labeled_img, check_ocr_box, get_caption_model_processor, get_yolo_model\n",
    "import torch\n",
    "from ultralytics import YOLO\n",
    "from PIL import Image\n",
    "device = 'cuda'\n",
    "model_path='weights/icon_detect_v1_5/model_v1_5.pt'\n",
    "\n",
    "som_model = get_yolo_model(model_path)\n",
    "\n",
    "som_model.to(device)\n",
    "print('model to {}'.format(device))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# two choices for caption model: fine-tuned blip2 or florence2\n",
    "import importlib\n",
    "import utils\n",
    "importlib.reload(utils)\n",
    "from utils import get_som_labeled_img, check_ocr_box, get_caption_model_processor, get_yolo_model\n",
    "# caption_model_processor = get_caption_model_processor(model_name=\"blip2\", model_name_or_path=\"weights/icon_caption_blip2\", device=device)\n",
    "caption_model_processor = get_caption_model_processor(model_name=\"florence2\", model_name_or_path=\"weights/icon_caption_florence\", device=device)\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "som_model.device, type(som_model) "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# reload utils\n",
    "import importlib\n",
    "import utils\n",
    "importlib.reload(utils)\n",
    "from utils import get_som_labeled_img, check_ocr_box, get_caption_model_processor, get_yolo_model\n",
    "\n",
    "# image_path = 'imgs/google_page.png'\n",
    "# image_path = 'imgs/windows_home.png'\n",
    "# image_path = 'imgs/windows_multitab.png'\n",
    "# image_path = 'imgs/omni3.jpg'\n",
    "# image_path = 'imgs/ios.png'\n",
    "# image_path = 'imgs/word.png'\n",
    "image_path = 'screenshot.png'\n",
    "# image_path = 'imgs/excel2.png'\n",
    "\n",
    "image = Image.open(image_path)\n",
    "image_rgb = image.convert('RGB')\n",
    "print('image size:', image.size)\n",
    "\n",
    "box_overlay_ratio = max(image.size) / 3200\n",
    "draw_bbox_config = {\n",
    "    'text_scale': 0.8 * box_overlay_ratio,\n",
    "    'text_thickness': max(int(2 * box_overlay_ratio), 1),\n",
    "    'text_padding': max(int(3 * box_overlay_ratio), 1),\n",
    "    'thickness': max(int(3 * box_overlay_ratio), 1),\n",
    "}\n",
    "BOX_TRESHOLD = 0.05\n",
    "\n",
    "import time\n",
    "start = time.time()\n",
    "ocr_bbox_rslt, is_goal_filtered = check_ocr_box(image_path, display_img = False, output_bb_format='xyxy', goal_filtering=None, easyocr_args={'paragraph': False, 'text_threshold':0.8}, use_paddleocr=True)\n",
    "text, ocr_bbox = ocr_bbox_rslt\n",
    "cur_time_ocr = time.time() \n",
    "\n",
    "dino_labled_img, label_coordinates, parsed_content_list = get_som_labeled_img(image_path, som_model, BOX_TRESHOLD = BOX_TRESHOLD, output_coord_in_ratio=True, ocr_bbox=ocr_bbox,draw_bbox_config=draw_bbox_config, caption_model_processor=caption_model_processor, ocr_text=text,use_local_semantics=True, iou_threshold=0.7, scale_img=False, batch_size=128)\n",
    "cur_time_caption = time.time() \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# plot dino_labled_img it is in base64\n",
    "import base64\n",
    "import matplotlib.pyplot as plt\n",
    "import io\n",
    "plt.figure(figsize=(15,15))\n",
    "\n",
    "image = Image.open(io.BytesIO(base64.b64decode(dino_labled_img)))\n",
    "plt.axis('off')\n",
    "\n",
    "plt.imshow(image)\n",
    "# print(len(parsed_content_list))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "df = pd.DataFrame(parsed_content_list)\n",
    "df['ID'] = range(len(df))\n",
    "\n",
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "parsed_content_list"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "df.to_csv(\"omni_parser.csv\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
